### ISSUE TITLE
TypeAliasType reference cycle errors depending on cycle declaration order

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#8879]
**Describe the bug**

I think this might be limited to a forward reference to a callable,
but mostly as the title describes, `type` keyword allows forward references
and both declarations contain forward references, so neither version should error.
It seems like when the callable is the forward reference the typed tuple becomes a
sequence of the union of all contained types instead of the ordered typed values.

**Code or Screenshots**

```python
from collections.abc import Callable

type FRetVal = tuple[bool, CFRefT]
type CFRefT = Callable[[], FRetVal]

# above and below type declarations differ only in order of declaration. Above causes error, below doesn't

type CRefT = Callable[[], RetVal]
type RetVal = tuple[bool, CRefT]

# no error for below types, doesn't include callable

type FRetNC = tuple[bool, FRef]
type FRef = tuple[str, FRetNC]

def fret_foo() -> FRetNC:
    ...

class Bar:
    state: bool
    fstate_incr: CFRefT
    state_incr: CRefT
    fref_val: FRef

    def use_ref(self):
        self.state, self.state_incr = self.state_incr()

    def use_fref(self):
        # Cannot assign to attribute "state" for class "Bar*"
        # Expression of type "bool | CFRefT" cannot be assigned to attribute "state" of class "Bar"
        # Cannot assign to attribute "fstate_incr" for class "Bar*"
        # Expression of type "bool | CFRefT" cannot be assigned to attribute "fstate_incr" of class "Bar"
        self.state, self.fstate_incr = self.fstate_incr()

    def use_val_fref(self):
        self.state, self.fref_val = fret_foo()
```

**VS Code extension or command-line**
pyright/pylance/pyright playground

**Comment by erictraut** [comment]
Thanks for the bug report. I appreciate the clear repro steps and the analysis.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.379.
